### ISSUE TITLE
[False positive] `avoid_redundant_argument_values` with dart defines

### ISSUE THREAD

**Issue by AlexV525** [dart-lang/linter#3447]
**Describe the issue**
Using the operator `!` for `bool.fromEnvironment` bring false positives.

**To Reproduce**
```dart
const bool someDefine = bool.fromEnvironment('someDefine');

void test({bool test = true}) {}

void main() {
  test(
    test: !someDefine, // <-- false positive
  );
}
```

**Comment by pq** [comment]
Thanks for this @AlexV525!

Looking at a fix w/ @bwilkerson, it seems that we want to guard against the case where an evaluated constant does not have a known value.  Reproducing now.

**Comment by AlexV525** [comment]
How can I determine which release contains the fix?

**Comment by pq** [comment]
Unfortunately I don't think a fix has landed.  IIRC this requires some upstream changes to the analyzer's constant evaluation.

**Comment by bwilkerson** [comment]
Can you remind me what changes are needed?

**Comment by pq** [comment]
@bwilkerson: there's a test (marked failing) in https://github.com/dart-lang/linter/pull/3449 that I think captures the expected behavior.

**Comment by bwilkerson** [comment]
Ok. As you noted above, it's likely a result of how the analyzer treats `fromEnvironment` values for the purpose of constant evaluation. Equality probably isn't defined right for this test.

**Comment by AlexV525** [comment]
Did we get this landed?

**Comment by pq** [comment]
Not yet. The constant evaluation work hasn't been tackled in analyzer yet AFAIK.

/fyi @scheglov @kallentu
